File: js/ball_class.js
define(["global_constants", "global_variables"], function (gameConstants, gameVariables) {
    var Ball = function () {
        // Set the ball to the center of the field and make it stopped at the beginning.
        this.x = gameConstants.HORIZONTAL_CENTER_OF_FIELD;
        this.y = gameConstants.VERTICAL_CENTER_OF_FIELD;
        this.horizontalVelocity = gameConstants.STOPPED;
        this.verticalVelocity = gameConstants.STOPPED;
        // Draw the ball with the proper positions and dimensions.
        this.draw = function () {
            gameConstants.CONTEXT.beginPath();
            gameConstants.CONTEXT.arc(this.x, this.y, gameConstants.BALL_RADIUS, gameConstants.BALL_START_ANGLE, gameConstants.BALL_END_ANGLE);
            gameConstants.CONTEXT.fill();
        };
        // Update the ball's position depending on velocity. Also do collision detection for the ball.
        this.updatePosition = function (rightPaddle, leftPaddle) {
            this.x += this.horizontalVelocity / gameConstants.FPS;
            this.y += this.verticalVelocity / gameConstants.FPS;
            // -------------------- Start of ball collision code --------------------
            
            // Top wall collision.
            if ((this.y - gameConstants.BALL_RADIUS) < gameConstants.TOP_WALL) {
                this.y = gameConstants.BALL_RADIUS;
                this.verticalVelocity = -this.verticalVelocity;
            }
            
            // Bottom wall collision.
            if ((this.y + gameConstants.BALL_RADIUS) > gameConstants.BOTTOM_WALL) {
                this.y = gameConstants.BOTTOM_WALL - gameConstants.BALL_RADIUS;
                this.verticalVelocity = -this.verticalVelocity;
            }
            
            // Right paddle collison on the front side of the paddle.
            if ((this.x + gameConstants.BALL_RADIUS) > rightPaddle.x) {
                if (this.y > rightPaddle.y) {
                    if (this.y < (rightPaddle.y + rightPaddle.height)) {
                        this.x = rightPaddle.x - gameConstants.BALL_RADIUS;
                        this.horizontalVelocity = -this.horizontalVelocity;
                    }
                }
            }
            
            // Right paddle collision on the top side of the paddle.
            if ((this.x + gameConstants.BALL_RADIUS) > rightPaddle.x) {
                if ((this.y + gameConstants.BALL_RADIUS) > rightPaddle.y) {
                    if ((this.y + gameConstants.BALL_RADIUS) < (rightPaddle.y + rightPaddle.height)) {
                        this.y = rightPaddle.y - gameConstants.BALL_RADIUS;
                        this.verticalVelocity = -this.verticalVelocity;
                    }
                }
            }
            
            // Right paddle collision on the bottom side of the paddle.
            if ((this.x + gameConstants.BALL_RADIUS) > rightPaddle.x) {
                if ((this.y - gameConstants.BALL_RADIUS) < (rightPaddle.y + rightPaddle.height)) {
                    if ((this.y - gameConstants.BALL_RADIUS) > rightPaddle.y) {
                        this.y = (rightPaddle.y + rightPaddle.height) + gameConstants.BALL_RADIUS;
                        this.verticalVelocity = -this.verticalVelocity;
                    }
                }
            }
            
            // Left paddle collision on the front side of the paddle.
            if ((this.x - gameConstants.BALL_RADIUS) < (leftPaddle.x + leftPaddle.width)) {
                if (this.y > leftPaddle.y) {
                    if (this.y < (leftPaddle.y + leftPaddle.height)) {
                        this.x = (leftPaddle.x + leftPaddle.width) + gameConstants.BALL_RADIUS;
                        this.horizontalVelocity = -this.horizontalVelocity;
                    }
                }
            }
            
            // Left paddle collision on the top side of the paddle.
            if ((this.x - gameConstants.BALL_RADIUS) < (leftPaddle.x + leftPaddle.width)) {
                if ((this.y + gameConstants.BALL_RADIUS) > leftPaddle.y) {
                    if ((this.y + gameConstants.BALL_RADIUS) < (leftPaddle.y + leftPaddle.height)) {
                        this.y = leftPaddle.y - gameConstants.BALL_RADIUS;
                        this.verticalVelocity = -this.verticalVelocity;
                    }
                }
            }
            
            // Left paddle collision on the bottom side of the paddle.
            if ((this.x - gameConstants.BALL_RADIUS) < (leftPaddle.x + leftPaddle.width)) {
                if ((this.y - gameConstants.BALL_RADIUS) < (leftPaddle.y + leftPaddle.height)) {
                    if ((this.y - gameConstants.BALL_RADIUS) > leftPaddle.y) {
                        this.y = (leftPaddle.y + leftPaddle.height) + gameConstants.BALL_RADIUS;
                        this.verticalVelocity = -this.verticalVelocity;
                    }
                }
            }
            
            // ******************** End of ball collision code ********************
            
            // Restart at center when the ball leaves the field and score a point for the paddle that scored.
            
            // When the ball goes past the right paddle...
            if ((this.x - gameConstants.BALL_RADIUS) > gameConstants.RIGHT_WALL) {
                // ... restart at the center of the field and mark a point for the left paddle.
                this.x = gameConstants.HORIZONTAL_CENTER_OF_FIELD;
                this.y = gameConstants.VERTICAL_CENTER_OF_FIELD;
                leftPaddle.score++;
            }
            
            // When the ball goes past the left paddle...
            if ((this.x + gameConstants.BALL_RADIUS) < gameConstants.LEFT_WALL) {
                // ... restart at the center of the field and mark a point for the right paddle.
                this.x = gameConstants.HORIZONTAL_CENTER_OF_FIELD;
                this.y = gameConstants.VERTICAL_CENTER_OF_FIELD;
                rightPaddle.score++;
            }
        };
    };
    return Ball;
});

File: rb/ball_class.rb
require_remote "rb/global_constants.rb"

class Ball
  
  # Make the instance variables accessible
  attr_accessor :x_pos
  attr_accessor :y_pos
  attr_accessor :horizontal_velocity
  attr_accessor :vertical_velocity
  
  def initialize
    @x_pos = GameConstants::HORIZONTAL_CENTER_OF_FIELD
    @y_pos = GameConstants::VERTICAL_CENTER_OF_FIELD
    @horizontal_velocity = GameConstants::STOPPED
    @vertical_velocity = GameConstants::STOPPED
  end
  
  def draw
    GameConstants::CONTEXT.fillStyle = "white"
    GameConstants::CONTEXT.beginPath
    GameConstants::CONTEXT.arc(@x_pos, @y_pos, GameConstants::BALL_RADIUS, GameConstants::BALL_START_ANGLE, GameConstants::BALL_END_ANGLE)
    GameConstants::CONTEXT.fill
  end
  
  def update_position (right_paddle, left_paddle)
    # Update position.
    @x_pos += @horizontal_velocity / GameConstants::FPS
    @y_pos += @vertical_velocity / GameConstants::FPS
    
    # -------------------- Start of ball collision code --------------------
    
    # Ball collision constants (these can't go into GameConstants because they use objects unknown to GameConstants, i.e., right_paddle and left_paddle)
    TOP_OF_BALL = @y_pos - GameConstants::BALL_RADIUS
    BOTTOM_OF_BALL = @y_pos + GameConstants::BALL_RADIUS
    RIGHT_SIDE_OF_BALL = @x_pos + GameConstants::BALL_RADIUS
    LEFT_SIDE_OF_BALL = @x_pos - GameConstants::BALL_RADIUS
    FRONT_SIDE_OF_RIGHT_PADDLE = right_paddle.x_pos
    TOP_OF_RIGHT_PADDLE = right_paddle.y_pos
    BOTTOM_OF_RIGHT_PADDLE = right_paddle.y_pos + right_paddle.height
    FRONT_SIDE_OF_LEFT_PADDLE = left_paddle.x_pos + left_paddle.width
    TOP_OF_LEFT_PADDLE = left_paddle.y_pos
    BOTTOM_OF_LEFT_PADDLE = left_paddle.y_pos + left_paddle.height
    
    # Top wall collision
    if TOP_OF_BALL < GameConstants::TOP_WALL
      @y_pos = GameConstants::TOP_WALL + GameConstants::BALL_RADIUS
      @vertical_velocity = -@vertical_velocity
    end
    
    # Bottom wall collision
    if BOTTOM_OF_BALL > GameConstants::BOTTOM_WALL
      @y_pos = GameConstants::BOTTOM_WALL - GameConstants::BALL_RADIUS
      @vertical_velocity = -@vertical_velocity
    end
    
    # Right paddle collision (when the ball goes past the front side of the paddle)
    if RIGHT_SIDE_OF_BALL > FRONT_SIDE_OF_RIGHT_PADDLE
      # Front side collision
      if @y_pos > TOP_OF_RIGHT_PADDLE && @y_pos < BOTTOM_OF_RIGHT_PADDLE
        @x_pos = FRONT_SIDE_OF_RIGHT_PADDLE - GameConstants::BALL_RADIUS
        @horizontal_velocity = -@horizontal_velocity
      end
      
      # Top side collision
      if BOTTOM_OF_BALL > TOP_OF_RIGHT_PADDLE && @y_pos < TOP_OF_RIGHT_PADDLE && BOTTOM_OF_BALL < BOTTOM_OF_RIGHT_PADDLE
        @y_pos = TOP_OF_RIGHT_PADDLE - GameConstants::BALL_RADIUS
        @vertical_velocity = -@vertical_velocity
      end
      
      # Bottom side collision
      if TOP_OF_BALL < BOTTOM_OF_RIGHT_PADDLE && @y_pos > BOTTOM_OF_RIGHT_PADDLE && TOP_OF_BALL > TOP_OF_RIGHT_PADDLE
        @y_pos = BOTTOM_OF_RIGHT_PADDLE + GameConstants::BALL_RADIUS
        @vertical_velocity = -@vertical_velocity
      end
    end
    
    # Left paddle collision (when the ball goes past the front side of the paddle)
    if LEFT_SIDE_OF_BALL < FRONT_SIDE_OF_LEFT_PADDLE
      # Front side collision
      if @y_pos > TOP_OF_LEFT_PADDLE && @y_pos < BOTTOM_OF_LEFT_PADDLE
        @x_pos = FRONT_SIDE_OF_LEFT_PADDLE + GameConstants::BALL_RADIUS
        @horizontal_velocity = -@horizontal_velocity
      end
      
      # Top side collision
      if BOTTOM_OF_BALL > TOP_OF_LEFT_PADDLE && @y_pos < TOP_OF_LEFT_PADDLE && BOTTOM_OF_BALL < BOTTOM_OF_LEFT_PADDLE
        @y_pos = TOP_OF_LEFT_PADDLE - GameConstants::BALL_RADIUS
        @vertical_velocity = -@vertical_velocity
      end
      
      # Bottom side collision
      if TOP_OF_BALL < BOTTOM_OF_LEFT_PADDLE && @y_pos > BOTTOM_OF_LEFT_PADDLE && TOP_OF_BALL > TOP_OF_LEFT_PADDLE
        @y_pos = BOTTOM_OF_LEFT_PADDLE + GameConstants::BALL_RADIUS
        @vertical_velocity = -@vertical_velocity
      end
      
      # ******************** End of ball collision code ********************
      
      # Restart at center when the ball leaves the ball leaves the field and score a point for the paddle that scored.
      
      # When the ball goes past the right paddle...
      if LEFT_SIDE_OF_BALL > GameConstants::RIGHT_WALL
        # ... restart at the center of the field and mark a point for the left paddle.
        @x_pos = GameConstants::HORIZONTAL_CENTER_OF_FIELD
        @y_pos = GameConstants::VERTICAL_CENTER_OF_FIELD
      end
      
      # When the ball goes past the left paddle...
      if RIGHT_SIDE_OF_BALL < GameConstants::LEFT_WALL
        # ... restart at the center of the field and mark a point for the right paddle.
        @x_pos = GameConstants::HORIZONTAL_CENTER_OF_FIELD
        @y_pos = GameConstants::VERTICAL_CENTER_OF_FIELD
      end
    end
  end
end